Classloader updates

dmatej opened the pull request:
### WHY

- GlassFish classloaders were quite obsoleted, still based on Java 1.4 features.
- Known bugs
- It was hard to read the code.
- The implementation was slow.
- Logging did not tell much.
- Tomcat WebappClassLoader sources diverged, however are somehow old too. Meanwhile GF implemented other things ....

### Fixed bugs along the path

* #23769 Obsoleted code, current JDK provides solution. JdbcLeakPrevention replaced.
* ACCAgentClassLoader implemented GlassfishUrlClassLoader which is not possible as it is used as a bootstrap classloader. Introduced in 15332ce2e69b14a910 .
* ReferenceCleaner (originally inside WebappClassLoader) did not work, now does.
  * Tried to reset enums (forbidden)
  * Tried to access fields without permission (forbidden)
  * Noisy logging when it ignored excluded fields. (removed, not interesting)

### Changes

* WebappClassLoader is now "parallel-capable". Means basically declaring that and using `getClassLoadingLock(name)` for locking.
* WebappClassLoader is stateful with an attempt to strictly control what can be called
  * URLClassLoaders are Closeable
  * originally was just started and not started, while it wasn't checked.
  * now can be NEW/RUNNING/CLOSED, however some methods can be called both when NEW and RUNNING, which is not good, but it would need much more work to simplify it (give all parameters to constructor and use the CL until it is closed).
* Using System.Logger instead of JUL.
* Removed redundant logs, added new logs, set these and watch:
    ```
    org.glassfish.web.loader.WebappClassLoader.level=FINEST
    org.glassfish.web.loader.JarFileManager.level=FINEST
    org.glassfish.web.loader.ReferenceCleaner.level=FINEST
    ```
* Several new package-visible classes split from WebappClassLoader, each responsible for some segment.
  * JarFileManager manages JAR files under classloader's control
      * synchronized blocks replaced by ReentrantReadWriteLock
      * JarFile management moved inside to a special scheduled task which closes files when they are not used. This was done also in original implementation, but managed by the StandardContext and WebappLoader.
      * JarFile instances not closed on startup "to avoid peaks" as they were immediately open again.     
      * Both period and the timeout are configurable by the JVM option. Original timeout was 90 seconds, now period is 15 , timeout 60 seconds.
  * ReferenceCleaner serves to resolve possible CL leaks. If there are ThreadLocal values forgotten, it should set them to null, static fields, etc. The implementation was already there, but did not work. 
  * RepositoryManager for work with repositories. Quite simple, but maybe in next iteration we could make it yet faster ;-)
* setJarPath removed - now the CL authoritatively defines the path; the path is hardcoded on several places and must not be changed.
* WebappLoader.createClassLoader doesn't start the classloader. The start split to own method. This makes also more visible the trick in the child ...
* 

### Performance

* Currently I don't have a JMH test with a large application, but tests we have are bit faster. It should depend on hardware, from my experience Intels don't have issues with synchronizations, but sparcs had. Not sure with current ARMs etc. Would be nice if someone cloud test it ... or if I would find time :-)

dmatej committed:
GlassfishBootstrapClassLoader is parallel capable

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
WebappClassLoader is parallel capable

- but it should be optimized.

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
WebappClassLoader - using System.Logger

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
GlassfishUrlClassLoader - logging

- Usage: to see order of things touched on start/stop

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Using Boolean.getBoolean instead of System.getProperty("...") != null

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
The stop method replaced by close, which is standard.

- Also removed the condition about starting.
- There is several reported issues related to the cleanup.

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Using constructor instead of foreach

- but it seems this constructor is not used at all.

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
WebappClassLoader usages - simplified

- using function references instead anonymous classes
- removed unused constructor

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Fixed obsoleted ResourceBundle cache cleanup

- No need for reflection any more

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Fixed issue #23769 - no need for reflection

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
ACCAgentClassLoader reverted back to URLClassLoader

- Usage of logging led to this exception:
  java.lang.IllegalStateException: getSystemClassLoader cannot be called
  during the system class loader instantiation

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
ACCClassLoader cleanup - using FileUtils

- Originally it used bit weird buffer size 8196 B
- try-with

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Fixed unit test compilation

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Issue #23769 Refactored deregistering JDBC drivers

- Old issues with this are not happening with current JDK
- See DriverManager code, changed in Java 8.

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
WebappClassLoader - protected fields made private

- loader doesn't have any children

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
WebappClassLoader - cleanup

- Reduced visibility of methods if they are not used outside
- Changed order of methods
- PreDestroy - annotation instead of implements
- FileUtils.copy and try-with
- Formatting
- getURI and getURL did the same thing

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
First step to WebappCl lifecycle control

- The CL is in fact stateful, but it is not controlled
- Simplified synchronization, using getClassLoadingLock
- Higher throughput, visible in tests
- However I have still doubts about reliability (no evidence)
- This commit passed Servlet, Concurrency and Faces TCKs

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Security manager changed to constant

- In the system it is set just once.

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Inlined init() and removed getResources()

- Now we see that system cl is final
- setResources can be called just before CL started (consistency)

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Cleanups

- notFoundResources is in fact a set
- @throws instead @exception
- removed copy pasted javadoc
- names of variables
- SM is already here

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Simplifying

- getDelegateClassLoader() as the code repeats, also removes double assignation
- resolveIfRequired(..) as the code repeats
- parent removed, is alerady in AbstractClass
  - that means also that it is not reset in close, but even before, it wasn't
    done in parent AbstractCL too, also this class cannot survive parent.
- The classloader is now final

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Added status checking to WebappClassLoader

- However V3 hacked original Apache code in the sake of new features. Since
  then the classloader is used even partially configured, we should resolve
  it later, but it will probably need larger refactoring.
- Even then we are able to move bit forward and separate methods used to
  configure the CL and methods used in standard traffic.
- Separated WebappLoader.startNestedClassLoader
  - symmetrical to stopNestedClassLoader
  - makes more visible when is the classloader started, or at least should be.
  - V3WebappLoader just uses already started classloader, but WebappLoader does
    further changes to it. To my surprise it doesn't cause issues ... yet.
- HINT: use org.glassfish.web.loader.WebappClassLoader.level=FINEST

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
JarFileManager split from WebappClassLoader

- All it's methods are synchronized, which is quite lame way of coding, but
  it can be simply revisited later
- No need for explicit custom locking in the classloader as this class is
  responsible for the whole segment.

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Safety - don't provide internal cache

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Repositories and files moved to RepositoryManager + optimizations

- Instead of two arrays now we have single list of Repository objects which
  can be transformed to list of RepositoryResource objects
- No need to enlarge jarFiles field when it is not used yet.
- Renamed resources to jndiResources because I confised them several times
  with repositories etc.

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Optimized JarFileManager

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Optimized JarFileManager, next gen.

- Using more advanced locking
- Moved responsibility for files to the manager.
- There should be done more work around StandardContext

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Quick maintenance in DeploymentContextImpl

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
The byteCodePreprocessors doesn't need to be concurrent list

- It is appended in the NEW phase, and used in RUNNING phase.

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
The libPath field was in fact a constant which must not change

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
All errors from toURL were ignored, so we can ignore them sooner

- ... and get better results
- MalformedUrlException should not even happen, am I right?

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
PathTimestamp list replaces the paths and lastModifiedDates arrays

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Changed method order, more logs

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
ReferenceCleaner split from WebappClassloader, named classloader

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Further cleanups

- Better javadocs, especially for ResourceEntry, Reloader
- createResourceEntry
- confusing File->URL conversions simplified
- entry.codebase is nullable just for classes
- StandardClassLoader - just Eclipse save actions
- GlassfishUrlClassLoader.toString now contains name of the classloader

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Fixed resource cleaning - based on tests

- Reproducer: anything using Arquillian (Concurrency TCK, Servlet TCK) plus
  hardcoded clearReferencesStatic set to true.
- Excluded enums
- setAccessible must be called for each getDeclaredFields

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Removed unused canonicalLoaderDir

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Final cleanups

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Transformers can be added later too.

- Reverted collection to original ConcurrentLinkedQueue

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

BalusC commented on appserver/appclient/client/acc/src/main/java/org/glassfish/appclient/client/acc/ACCClassLoader.java:
```diff
@@ -187,27 +181,15 @@ private Class<?> findClassUnshadowed(String name) throws ClassNotFoundException
+                throw new ClassNotFoundException(className);
             }
+            ByteArrayOutputStream baos = new ByteArrayOutputStream();
+            FileUtils.copy(is, baos, is.available());
```
3rd argument represents hopefully the buffer size not the size to copy? (this FileUtils doesn't seem to be from Apache Commons so I couldn't verify).

dmatej replied:
FileUtils attracted my attention in my second workspace, here it uses NIO and the third is used as a size.
`        outChannel.transferFrom(inChannel, 0, size);`

>  The maximum number of bytes to be transferred; must be non-negative

Maybe would be better to simply use Long.MAX_VALUE ...?

BalusC replied:
Probably.

This just caught my attention because `is.available()` ranks among the top of beginner's misunderstandings.

dmatej replied:
To summarize: 
* in.available is correct here, because the input is a file and the third argument is a max count of bytes to copy.
* In next PR I will review FileUtils and document these methods, and probably explicitly warn to avoid using Long.MAX_VALUE, and also add loop to be sure we read all we expected, because based on FileChannel javadoc it is not guaranteed to be done in a single step.

BalusC commented on appserver/web/war-util/src/main/java/org/glassfish/web/loader/RepositoryManager.java:
```diff
@@ -0,0 +1,113 @@
+    private static class Repository {
+
+        public String name;
+        public File directory;
```
I would have made these `final` for clarity.

BalusC commented on appserver/web/war-util/src/main/java/org/glassfish/web/loader/RepositoryManager.java:
```diff
@@ -0,0 +1,113 @@
+    static class RepositoryResource {
+
+        public String name;
+        public File file;
```
I would have made these `final` for clarity.

BalusC commented on appserver/web/war-util/src/main/java/org/glassfish/web/loader/ResourceEntry.java:
```diff
@@ -38,40 +48,117 @@ public class ResourceEntry {
+     */
+    ResourceEntry(URL url) {
+        codeBase = url;
+        source = url;
```
I'd have used `this(url, url)`.

BalusC commented on appserver/web/war-util/src/main/java/org/glassfish/web/loader/ResourceEntry.java:
```diff
@@ -38,40 +48,117 @@ public class ResourceEntry {
+     */
+    ResourceEntry(URL codeBase, URL source) {
+        this.codeBase = codeBase;
+        this.source = source;
```
These two props should probably be marked `final`?

dmatej replied:
Nope, this part is ugly (still as in original). See javadocs; some attributes are reset when the class is loaded.
Perhaps would be better to have three types of entries: UnloadedClass, LoadedClass and Resource, but finally I stopped here. It would not resolve much, maybe objects would be bit smaller ... maybe we could separate classes and resources, hmm ... maybe next time :-)

BalusC replied:
Ew.

BalusC commented on appserver/web/war-util/src/main/java/org/glassfish/web/loader/JarFileManager.java:
```diff
@@ -0,0 +1,362 @@
+import static java.lang.System.Logger.Level.TRACE;
+import static java.lang.System.Logger.Level.WARNING;
+import static java.util.concurrent.Executors.newScheduledThreadPool;
+import static org.glassfish.web.loader.LogFacade.getString;
```
Static import lines should be in top.

dmatej replied:
> Static import lines should be in top.

This is not defined on GF yet, most classes have them closer to the impl (I prefer this), less at the top and some in between (which is the most ugly).

BalusC replied:
It violates standard conventions and default rules of a lot of static code analyzers (sonar, checkstyle, etc).

dmatej replied:
I believe that depends on configuration, there is no such global standard.

I use SonarQube and this is not seen as violation. I even tried to find such rule, but to my surprise I couldn't. I know it is supported in checkstyle and I too thought it is in SonarQube too.

Btw pretty much work to do with GF. I couldn't run it for a while, because Sonar analysis crashes for OOME now :-(
https://sonarcloud.io/project/overview?id=dmatej_glassfish

BalusC replied:
Okay.

dmatej replied:
To avoid future discussions and conflicts I started trivial survey ... at least it could be representative between friends of GlassFish :-)
Based on that I will probably introduce rule in Checkstyle and fix all classes.
https://twitter.com/dmatej/status/1614963057842688000

BalusC commented on nucleus/deployment/common/src/main/java/org/glassfish/deployment/common/DeploymentContextImpl.java:
```diff
@@ -58,7 +53,12 @@
+import static java.lang.Boolean.FALSE;
+import static java.lang.Boolean.TRUE;
+import static java.util.logging.Level.FINE;
+import static java.util.logging.Level.FINEST;
```
Static import lines should remain in top.

dmatej committed:
Reflecting BalusC review - finals, better constructor

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

hs536 commented on appserver/web/war-util/src/main/java/org/glassfish/web/loader/WebappClassLoader.java:
```diff
@@ -179,7 +179,7 @@ public class WebappClassLoader extends GlassfishUrlClassLoader
      * loaded by this classloader.
      */
-    private final boolean packageDefinitionEnabled = System.getProperty("package.definition") != null;
+    private final boolean packageDefinitionEnabled = Boolean.getBoolean("package.definition");
```
Isn't the value obtained using "package.definition" a list of package names, not a boolean string?

dmatej replied:
Nice catch, I tried to search the string and it seems to me it is even worse for a long time ... the property is set by catalina not to System properties, but via java.security.Security.setProperty. So this was always false.

pzygielo replied:
Nice finding. I believe it deserves separate issue, that could be solved separately in dedicated PR.

dmatej replied:
I brought back the original System.getProperty. Even if it is not probably worth of merging to older branches and SecurityManager will die soon, GF should be correct at the moment, so after merging this I can create another tiny PR :)

dmatej committed:
Reflecting Hiroki's review - reverted package.definition system property.

- note: it is all wrong for a long time, the property is set via
  Security.setProperty, not System.setProperty.

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

arjantijms commented on appserver/web/war-util/src/main/java/com/sun/enterprise/glassfish/web/WarHandler.java:
```diff
@@ -263,10 +262,10 @@ protected void configureLoaderProperties(WebappClassLoader cloader, WebXmlParser
                 }
             } catch (final Exception e) {
-                // Catch and ignore any exception in case the JAR file is empty.
+                LOG.log(Level.FINEST, "Could not add file " + file, e);
```
No need to change, but just a remark: many places use a static import for FINEST.

dmatej replied:
True, but not in WarHandler, I checked it now for sure. I tried to limit changes outside of loader, despite I think there is one or two classes where I "failed" and committed some eclipse autosave changes (foreach, override, braces).

arjantijms commented on appserver/web/war-util/src/main/java/org/glassfish/web/loader/JarFileManager.java:
```diff
@@ -0,0 +1,362 @@
+            closeJarFiles(files);
+        } finally {
+            // No need to interrupt, just cancel next executions
+            this.unusedJarsCheck.cancel(false);
```
> this.unusedJarsCheck

Is there another "unusedJarsCheck" in scope that would be invoked otherwise?
